File: app/lib/typography.rb
# Applies SmartyPants typography to a social post: curly quotes, a real ellipsis, and en and em
# dashes, from the characters someone types on a keyboard.
#
# This is the same typography as the blog, through Formattable#smartypants. A post and the entry it
# links to must not use a different apostrophe, so don't write a second set of rules here.
#
# A URL passes through unchanged, and that is the whole reason this class exists instead of calling
# the helper directly. SmartyPants reads the characters of an address as punctuation:
# `example.com/a--b` becomes an en dash, `/a...b` becomes an ellipsis, and `?q="a"` becomes curly
# quotes. Each one is a dead link, and nothing reports it.
class Typography
  # Extending the helper is what keeps this class from holding a rule of its own.
  extend Formattable

  # What each address becomes while SmartyPants reads the text. U+FFFC is OBJECT REPLACEMENT
  # CHARACTER, which is exactly what it says: one character standing in for something that isn't
  # text.
  #
  # The mask is necessary and splitting the text at each address is not enough. SmartyPants decides
  # which way a quote curls from the characters on either side of it, so it has to read the whole
  # sentence at once. Split, `He said "see <link> now"` opens the quote and never closes it,
  # because the two marks land in different pieces.
  PLACEHOLDER = "￼".freeze

  class << self
    # @param text [String, nil] the text as the author wrote it.
    # @return [String] the same text, with the site's typography applied everywhere but in a URL.
    def apply(text)
      # The text itself can't be allowed to hold the mask, or the addresses would go back in the
      # wrong places. That character isn't text and renders as nothing, so this loses no words.
      text = text.to_s.delete(PLACEHOLDER)
      return text if text.empty?

      urls = []
      converted = convert(mask(text, urls))

      # Hand back the text with no typography at all if a mask went missing. A straight apostrophe
      # is a small thing; an address that another address replaced is a link to the wrong page.
      return text unless converted.count(PLACEHOLDER) == urls.length

      index = -1
      converted.gsub(PLACEHOLDER) { urls[index += 1] }
    end

    private

    # Writes the text with one mask in place of each address, and collects those addresses.
    #
    # @param text [String] the text to mask.
    # @param urls [Array<String>] filled in, in the order the addresses appear.
    # @return [String] the masked text.
    def mask(text, urls)
      masked = +""
      last = 0

      masked_ranges(text).each do |range|
        masked << text[last...range.begin] << PLACEHOLDER
        urls << text[range]
        last = range.end
      end

      masked << text[last..].to_s
    end

    # SmartyPants writes HTML entities, because it's an HTML renderer: it gives `&rsquo;`, not `’`.
    # A post holds characters, so the decode is necessary, not decoration.
    #
    # @param chunk [String] the masked text.
    # @return [String] the text with typography applied and entities decoded.
    def convert(chunk)
      return chunk if chunk.empty?

      HTMLEntities.new.decode(smartypants(chunk).to_s)
    end

    # Every address and every mention, in order, with no overlaps: a mention that starts inside an
    # address is part of that address and doesn't get a mask of its own.
    #
    # Mentions are masked because an IDN handle starts with `xn--`, which SmartyPants would turn
    # into an en dash.
    #
    # @param text [String] the text to scan.
    # @return [Array<Range>] character ranges to mask, in order.
    def masked_ranges(text)
      urls = SocialText.url_ranges(text)
      tokens = []
      text.scan(Bluesky::MENTION_PATTERN) do
        start_char, end_char = Regexp.last_match.offset(1)
        tokens << (start_char...end_char) unless urls.any? { |range| range.cover?(start_char) }
      end
      (urls + tokens).sort_by(&:begin)
    end
  end
end

File: app/frontend/lib/typography.js
// The part of `Typography` that changes a LENGTH, in the browser.
//
// ⚠️ It writes no quotation mark, and that is correct: `"` becomes `“` and `'` becomes `’`, one
// character in place of one, so the direction of a quote cannot change a count. Deciding that
// direction is the hard half of SmartyPants and the half no small copy could match — and it does
// not have to be matched.
//
// ⚠️ **This file is for the COUNT and must never render text a person reads.** It would give a
// post with straight quotes.
//
// ⚠️ A URL passes through untouched, for the same reason as in the Ruby: `example.com/a--b` would
// otherwise become an en dash, and that link is dead.

// The bare-URL rule of `Bluesky::URL_PATTERN`: everything up to whitespace, with `trimUrl` then
// deciding where the address actually ends.
const URL = /(?:^|[$|\W])(https?:\/\/\S+)/g;

// `Bluesky::MENTION_PATTERN`. ⚠️ A mention is masked for the same reason as an address: an IDN
// handle starts with `xn--`, which the dash rule would turn into an en dash and count one short.
const MENTION = /(?:^|[$|\W])(@(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)/g;

// `Bluesky::URL_TERMINAL` and `Bluesky::URL_WRAPPERS`.
const URL_TERMINAL = /[\p{Alphabetic}\p{Nd}\-_~/#@$&*+=%]/u;
const URL_WRAPPERS = { ')': '(', ']': '[', '>': '<', '}': '{' };

const countOf = (text, char) => text.split(char).length - 1;

/**
 * Whether an address can stop at its last character.
 * @param {string} url - The candidate address.
 * @returns {boolean} True when the last character belongs to the address.
 */
function endsHere (url) {
  const last = url.at(-1);
  if (URL_TERMINAL.test(last)) return true;

  // A closing bracket belongs to the address only when the address opened it.
  const opener = URL_WRAPPERS[last];
  return Boolean(opener) && countOf(url, opener) >= countOf(url, last);
}

/**
 * Removes whatever at the end of a match belongs to the sentence rather than the address.
 * @param {string} url - The address as matched.
 * @returns {string} The address alone.
 */
function trimUrl (url) {
  let out = String(url);
  while (out !== '' && !endsHere(out)) out = out.slice(0, -1);
  return out;
}

// One character standing in for an address while the rules run. U+FFFC is OBJECT REPLACEMENT
// CHARACTER, and it is one grapheme, so a mask can't change a count either.
const PLACEHOLDER = '￼';

// Longest first: four dashes give "—-" and five give "—–".
const RULES = [
  [/\.\.\./g, '…'],
  [/\. \. \./g, '…'],
  [/---/g, '—'],
  [/--/g, '–']
];

/**
 * Applies the typography rules that change a character count.
 * @param {string} text - The text as the author wrote it.
 * @returns {string} The text with ellipses and dashes collapsed, addresses untouched.
 */
export function apply (text) {
  // The text itself must not hold the mask, or the addresses would go back in the wrong places.
  let source = String(text ?? '').replaceAll(PLACEHOLDER, '');
  if (source === '') return source;

  const masked = [];
  URL.lastIndex = 0;
  source = source.replace(URL, (whole, url) => {
    const trimmed = trimUrl(url);
    if (trimmed === '') return whole;

    masked.push(trimmed);
    // Keep whatever came before the address, and whatever the trim took off the end: the pattern
    // consumes one character of boundary, and the punctuation still needs its typography.
    return whole.slice(0, whole.length - url.length) + PLACEHOLDER + url.slice(trimmed.length);
  });

  // Mentions are masked after the addresses, so a handle inside a URL is already covered.
  MENTION.lastIndex = 0;
  source = source.replace(MENTION, (whole, handle) => {
    masked.push(handle);
    return whole.slice(0, whole.length - handle.length) + PLACEHOLDER;
  });

  let converted = source;
  for (const [pattern, replacement] of RULES) converted = converted.replace(pattern, replacement);

  // Hand back the text unchanged if a mask went missing, as the Ruby does.
  let index = -1;
  const masks = converted.split(PLACEHOLDER).length - 1;
  if (masks !== masked.length) return String(text ?? '');

  return converted.replaceAll(PLACEHOLDER, () => masked[++index]);
}
